amethyst_ui/src/button.rs

```diff
@@ -0,0 +1,224 @@
+    /// etc. as well.
+    /// Use [`with_text`](#with_text) to just change the underlying text.
+    pub fn with_uitext(mut self, text: UiText) -> Self {
+        self.text = text;
+        self
+    }
+
+    /// This will set the rendered characters within the button. Use this to just change what
+    /// characters will appear. If you need to change the font size, color, etc., then you should
+    /// use
+    /// [`with_uitext`](#with_uitext) and provide a new `UiText` object.
+    pub fn with_text<S>(mut self, text: S) -> Self
+    where
+        S: ToString,
+    {
+        self.text.text = text.to_string();
+        self
+    }
+
+    /// Replace the default UiImage with `image`.
+    pub fn with_image(mut self, image: UiImage) -> Self {
+        self.image = image;
+        self
+    }
+
+    /// Use a different font for the button text.
+    pub fn with_font(mut self, font: FontHandle) -> Self {
+        self.text.font = font;
+        self
+    }
+
+    pub fn with_position(mut self, x: f32, y: f32) -> Self {
+        self.transform = if let Some(mut t) = self.transform.take() {
+            t.local_x = x;
+            t.global_x = x;
+            t.local_y = y;
+            t.global_y = y;
+            Some(t)
+        } else {
+            let mut id = self.name.to_string();
+            id.push_str("_new_transform");
+            Some(UiTransform::new(
+                id,
+                x,
+                y,
+                DEFAULT_Z,
+                DEFAULT_WIDTH,
+                DEFAULT_HEIGHT,
+                DEFAULT_TAB_ORDER))
+        };
+        self
+    }
+
+    fn build_text(&mut self, image: &Entity, world: &mut World) -> Entity {
+        let mut id = self.name.to_string();
+        id.push_str("_btn_txt");
+        world
+            .create_entity()
+            .with(UiTransform::new(id, 0., 0., -1., 0., 0., 10))
+            .with(Anchored::new(Anchor::Middle))
+            .with(Stretched::new(Stretch::XY, 0., 0.))
+            .with(self.text.clone())
+            .with(Parent {
+                entity: image.clone(),
+            })
+            .build()
+    }
+
+    fn build_image(&mut self, world: &mut World) -> Entity {
+        let mut image_builder = world.create_entity().with(self.image.clone());
+        if let Some(parent) = self.parent.take() {
+            image_builder = image_builder.with(parent);
+        }
+        if let Some(transform) = self.transform.take() {
+            image_builder = image_builder.with(transform);
+        }
+        if let Some(anchored) = self.anchored.take() {
+            image_builder = image_builder.with(anchored);
+        }
+        if let Some(stretched) = self.stretched.take() {
+            image_builder = image_builder.with(stretched);
+        }
+
+        image_builder.build()
+    }
+
+    /// Create the UiButton based on provided configuration parameters.
+    pub fn build(mut self, world: &mut World) -> UiButton {
```

Comment:
Make another of the "Resources" structs, that contain the storages required for creating a button (WriteStorage<'a, UiImage> etc), and make the build function take that resources struct. Make the resources struct contain `Entities<'a>` so you can create entities.

Reply:
How do I create replicate the current `World` centric functionality with Storages and Entities? Does this new Resources struct also need to derive SystemData?